tests/robustness/failpoint_triggers.go

```diff
@@ -0,0 +1,195 @@
+	"fmt"
+	"strings"
+	"testing"
+	"time"
+
+	"go.uber.org/zap"
+
+	"go.etcd.io/etcd/api/v3/version"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.etcd.io/etcd/tests/v3/framework/e2e"
+)
+
+type trigger interface {
+	Trigger(t *testing.T, ctx context.Context, member e2e.EtcdProcess, clus *e2e.EtcdProcessCluster) error
+	AvailabilityChecker
+}
+
+type triggerDefrag struct{}
+
+func (t triggerDefrag) Trigger(_ *testing.T, ctx context.Context, member e2e.EtcdProcess, _ *e2e.EtcdProcessCluster) error {
+	cc, err := clientv3.New(clientv3.Config{
+		Endpoints:            member.EndpointsGRPC(),
+		Logger:               zap.NewNop(),
+		DialKeepAliveTime:    1 * time.Millisecond,
+		DialKeepAliveTimeout: 5 * time.Millisecond,
+	})
+	if err != nil {
+		return fmt.Errorf("failed creating client: %w", err)
+	}
+	defer cc.Close()
+	_, err = cc.Defragment(ctx, member.EndpointsGRPC()[0])
+	if err != nil && !strings.Contains(err.Error(), "error reading from server: EOF") {
+		return err
+	}
+	return nil
+}
+
+func (t triggerDefrag) Available(e2e.EtcdProcessClusterConfig, e2e.EtcdProcess) bool {
+	return true
+}
+
+type triggerCompact struct{}
+
+func (t triggerCompact) Trigger(_ *testing.T, ctx context.Context, member e2e.EtcdProcess, _ *e2e.EtcdProcessCluster) error {
+	cc, err := clientv3.New(clientv3.Config{
+		Endpoints:            member.EndpointsGRPC(),
+		Logger:               zap.NewNop(),
+		DialKeepAliveTime:    1 * time.Millisecond,
+		DialKeepAliveTimeout: 5 * time.Millisecond,
+	})
+	if err != nil {
+		return fmt.Errorf("failed creating client: %w", err)
+	}
+	defer cc.Close()
+	resp, err := cc.Get(ctx, "/")
+	if err != nil {
+		return err
+	}
+	_, err = cc.Compact(ctx, resp.Header.Revision)
+	if err != nil && !strings.Contains(err.Error(), "error reading from server: EOF") {
+		return err
+	}
+	return nil
+}
+
+func (t triggerCompact) Available(e2e.EtcdProcessClusterConfig, e2e.EtcdProcess) bool {
+	return true
+}
+
+type triggerBlackhole struct {
+	waitTillSnapshot bool
+}
+
+func (tb triggerBlackhole) Trigger(t *testing.T, ctx context.Context, member e2e.EtcdProcess, clus *e2e.EtcdProcessCluster) error {
+	return blackhole(t, ctx, member, clus, tb.waitTillSnapshot)
+}
+
+func (tb triggerBlackhole) Available(config e2e.EtcdProcessClusterConfig, process e2e.EtcdProcess) bool {
+	v, err := e2e.GetVersionFromBinary(e2e.BinPath.Etcd)
+	if err != nil {
+		panic(err)
+	}
+	// TODO: Deflake waiting for snapshot for v3.4.X
+	if tb.waitTillSnapshot && v.LessThan(version.V3_5) {
+		return false
+	}
+	return config.ClusterSize > 1 && process.PeerProxy() != nil
+}
+
+func blackhole(t *testing.T, ctx context.Context, member e2e.EtcdProcess, clus *e2e.EtcdProcessCluster, shouldWaitTillSnapshot bool) error {
+	proxy := member.PeerProxy()
+
+	// Blackholing will cause peers to not be able to use streamWriters registered with member
+	// but peer traffic is still possible because member has 'pipeline' with peers
+	// TODO: find a way to stop all traffic
```

Comment:
Just curious, how `member has 'pipeline' with peers` allows peer traffic given proxy will drop any packets? 

I thought pipeline will keep the msg in the buffer and throws error like `dropped internal Raft message since sending buffer is full (overloaded network)` if the remote peer is not reachable.

Reply:
I didn't investigate, I think this is something that @lavacat noticed.